### ISSUE TITLE
Lens when composed seem to fail to support type changing

### ISSUE THREAD

**Issue by jchia** [kcsongor/generic-lens#24]
When composing two `field` or `position` `Lens`es to obtain a type-changing `Lens`, the compiler generates an error. It seems that the error shows up when the value being changed is a record (not a tuple) and the result of the composition is not a simple `Lens' s t`.

For details, see test case and compiler errors below:

**Test case:**
```
{-# LANGUAGE DataKinds, DeriveGeneric, TypeApplications #-}

import Control.Lens
import Data.Generics.Product.Fields
import Data.Generics.Product.Positions
import GHC.Generics

data Foo a b = Foo { x1 :: a, x2 :: b } deriving (Generic, Show)

data Bar a b = Bar { x3 :: Foo a b, x4 :: Int } deriving (Generic, Show)

tup :: ((Int, Char), Int)
tup = ((1, 'a'), 2)
tup2, tup3, tup4 :: ((Char, Char), Int)
tup2 = tup & _1 . _1 %~ toEnum  -- Works.
tup3 = tup & x %~ toEnum  -- Works also with type annotation.
  where x :: Lens ((Int, Char), Int) ((Char, Char), Int) Int Char
        x = _1 . _1
-- Works.
tup4 = tup & position @1 . position @1 %~ toEnum

foo :: Foo Int Char
foo = Foo 1 'a'
foo2, foo3 :: Foo Char Char
foo2 = foo & field @"x1" %~ toEnum  -- Works when there's just one 'field'.
foo3 = foo & position @1 %~ toEnum -- Works when there's just one 'position'.

bar :: Bar Int Char
bar = Bar (Foo 1 'a') 2
bar2, bar3, bar4 :: Bar Char Char
-- Doesn't work, error at first 'field' (Couldn't match type ‘Int’ with ‘Char’ arising from a use of ‘field’).
bar2 = bar & field @"x3" . field @"x1" %~ toEnum
-- Type annotation doesn't help.
bar3 = bar & l %~ toEnum
  where l :: Lens (Bar Int Char) (Bar Char Char) Int Char
        l = field @"x3" . field @"x1"
-- Doesn't work, error at first 'position' (Couldn't match type ‘Int’ with ‘Char’ arising from a use of ‘position’).
bar4 = bar & position @1 . position @1 %~ toEnum
-- Works if we stick to simple Lens' (modify to the same type).
bar5 :: Bar Int Char
bar5 = bar & field @"x3" . field @"x1" %~ (+1)

main :: IO ()
main = print bar5
```

**Compiler errors:**
```
/home/jchia/hs/test/app/Main3.hs:32:14: error:
    • Couldn't match type ‘Int’ with ‘Char’
        arising from a use of ‘field’
    • In the first argument of ‘(.)’, namely ‘field @"x3"’
      In the first argument of ‘(%~)’, namely ‘field @"x3" . field @"x1"’
      In the second argument of ‘(&)’, namely
        ‘field @"x3" . field @"x1" %~ toEnum’
   |
32 | bar2 = bar & field @"x3" . field @"x1" %~ toEnum
   |              ^^^^^^^^^^^

/home/jchia/hs/test/app/Main3.hs:36:13: error:
    • Couldn't match type ‘Int’ with ‘Char’
        arising from a use of ‘field’
    • In the first argument of ‘(.)’, namely ‘field @"x3"’
      In the expression: field @"x3" . field @"x1"
      In an equation for ‘l’: l = field @"x3" . field @"x1"
   |
36 |         l = field @"x3" . field @"x1"
   |             ^^^^^^^^^^^

/home/jchia/hs/test/app/Main3.hs:36:27: error:
    • Couldn't match type ‘Char’ with ‘Int’
        arising from a use of ‘field’
    • In the second argument of ‘(.)’, namely ‘field @"x1"’
      In the expression: field @"x3" . field @"x1"
      In an equation for ‘l’: l = field @"x3" . field @"x1"
   |
36 |         l = field @"x3" . field @"x1"
   |                           ^^^^^^^^^^^

/home/jchia/hs/test/app/Main3.hs:38:14: error:
    • Couldn't match type ‘Int’ with ‘Char’
        arising from a use of ‘position’
    • In the first argument of ‘(.)’, namely ‘position @1’
      In the first argument of ‘(%~)’, namely ‘position @1 . position @1’
      In the second argument of ‘(&)’, namely
        ‘position @1 . position @1 %~ toEnum’
   |
38 | bar4 = bar & position @1 . position @1 %~ toEnum
   |              ^^^^^^^^^^^
```

**Comment by kcsongor** [comment]
Thanks for reporting this, currently type-changing lenses only work when exactly 1 type parameter is involved in the change. I’m currently working on fixing this as part of the solution to https://github.com/kcsongor/generic-lens/issues/23.

So in your case, the issue is that the x3 field mentions both type parameters, and this trips up the inference algorithm.

**Comment by kcsongor** [comment]
As I thought, the fix for #23 also fixed this. I'll add your test case to the test suite and make a release later today.
Thanks!

**Comment by jchia** [comment]
Thanks for working on such a useful library and fixing this issue.
